presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,424 @@
+            if (rightResult.cost.hasUnknownComponents()) {
+                return UNKNOWN_COST_RESULT;
+            }
+            if (rightResult.cost.equals(INFINITE_COST)) {
+                return INFINITE_COST_RESULT;
+            }
+            PlanNode right = rightResult.planNode.orElseThrow(() -> new IllegalStateException("no planNode present"));
+
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> sortedOutputSymbols = Stream.concat(left.getOutputSymbols().stream(), right.getOutputSymbols().stream())
+                    .filter(outputSymbols::contains)
+                    .collect(toImmutableList());
+
+            // Cross joins can't filter symbols as part of the join
+            // If we're doing a cross join, use all output symbols from the inputs and add a project node
+            // on top
+            List<Symbol> joinOutputSymbols = sortedOutputSymbols;
+            if (joinConditions.isEmpty() && joinFilters.isEmpty()) {
+                joinOutputSymbols = Stream.concat(left.getOutputSymbols().stream(), right.getOutputSymbols().stream())
+                        .collect(toImmutableList());
+            }
+
+            JoinEnumerationResult result = setJoinNodeProperties(new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    joinConditions,
+                    joinOutputSymbols,
+                    joinFilters.isEmpty() ? Optional.empty() : Optional.of(and(joinFilters)),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty()));
+
+            if (!joinOutputSymbols.equals(sortedOutputSymbols)) {
+                PlanNode resultNode = new ProjectNode(idAllocator.getNextId(), result.planNode.get(), identity(sortedOutputSymbols));
+                result = new JoinEnumerationResult(costProvider.getCumulativeCost(resultNode), Optional.of(resultNode));
+            }
+
+            return result;
+        }
+
+        private JoinEnumerationResult getJoinSource(List<PlanNode> nodes, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                ImmutableList.Builder<Expression> predicates = ImmutableList.builder();
+                predicates.addAll(allInference.generateEqualitiesPartitionedBy(outputSymbols::contains).getScopeEqualities());
+                stream(EqualityInference.nonInferrableConjuncts(allFilter))
+                        .map(conjuct -> allInference.rewriteExpression(conjuct, outputSymbols::contains))
+                        .filter(Objects::nonNull)
+                        .forEach(predicates::add);
+                Expression filter = combineConjuncts(predicates.build());
+                if (!(TRUE_LITERAL).equals(filter)) {
```

Comment:
nit: redundant parentheses around TRUE_LITERAL

`if (!TRUE_LITERAL.equals(filter))` should be sufficient